### PR TITLE
Added c++14 and c++17 standards.

### PR DESCRIPTION
**Pull request by TurkeyMan**

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Auto-merge toggled on

**Review by WalterBright — APPROVED** [review]
Do these have any effect on anything?

**Comment by thewilsonator** [comment]
No, they are just constants for code to use to define compatibilities at the moment, i.e. the C++ compiler equivalent of `#define __cplusplus whatever` for the appropriate value of `whatever`. 

The switch will eventually control mangling and another oddities of C++ versioning.

**Comment by jacob-carlborg** [comment]
Not all the branches are being tested. But if you don’t think that’s important.

**Comment by TurkeyMan** [comment]
Tests will appear for usages of these enums shortly...

File: src/dmd/mars.d [hunk 1]
```diff
@@ -1653,6 +1653,10 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
                 params.cplusplus = CppStdRevision.cpp98;
             else if (strcmp(p + 8, "c++11") == 0)
                 params.cplusplus = CppStdRevision.cpp11;
+            else if (strcmp(p + 8, "c++14") == 0)
+                params.cplusplus = CppStdRevision.cpp14;
+            else if (strcmp(p + 8, "c++17") == 0)
+                params.cplusplus = CppStdRevision.cpp17;
```
Comment (wilzbach) [inline src/dmd/mars.d]: Would be great to implement a usage subpage that lists all available options like it's done for e.g. `-transition`

**Comment by jacob-carlborg** [comment]
C++ compilers (at least Clang and GCC) allow to specify dialects of the language it supports. For example, gnu++17. Is that something that impacts us?

**Comment by thewilsonator** [comment]
No see the PR for the implementation of this feature.

**Comment by TurkeyMan** [comment]
I don't think it affects us. As far as I know, those affect what syntax is accepted by the C++ compiler, and has no affect on ABI, which is all we are worried about.

**Comment by TurkeyMan** [comment]
It would be great to put `CppStdRevision` in druntime... but where?

**Comment by jacob-carlborg** [comment]
> I don't think it affects us. As far as I know, those affect what syntax is accepted by the C++ compiler, and has no affect on ABI, which is all we are worried about.

It's not just syntax, it adds new features. Not sure if it changes any existing ones that could affect the ABI.

**Comment by jacob-carlborg** [comment]
> It would be great to put CppStdRevision in druntime... but where?

Do you think that would be used a lot by end users?

**Comment by TurkeyMan** [comment]
Probably not... unless they're binding to their own C++ code which is version sensitive.
I mean, anywhere that `__traits(getTargetInfo, "cppStd")` is used, then you need that enum.

**Comment by WalterBright** [comment]
1. `__traits(getTargetInfo, "cppStd");`
2. `core.stdcpp`
3. `-stdc++=`
4. `c++14`

Not very predictable.

**Comment by TurkeyMan** [comment]
Well, they're kinda different stuff. `"cppStd"` is to be read "C++ standard", that is, the C++ language standard.
`core.stdcpp` refers to the C++ `std` library.
I agree `-stdc++=` is not great; I suggested `-cppstd=`, such that it matches with #1.
`c++14` is a language standard key as defined by C++, and we have no influence over this keyspace.

The only thing I'd change from your list is `-stdc++=` to `-cppstd=` (as I suggested from the start) to match with #1, since those 2 items are indeed strongly related.

**Comment by jacob-carlborg** [comment]
I think we should use `c++` everywhere it's possible. It's not possible in a module name, hence `cpp`.

File: src/dmd/mars.d (hunk 1 above)
Comment (wilzbach) [reply src/dmd/mars.d]: -> https://github.com/dlang/dmd/pull/9217

**Comment by wilzbach** [comment]
No one noticed that the CLI docs weren't updated -> fix: https://github.com/dlang/dmd/pull/9217#discussion_r245809447

**Comment by WalterBright** [comment]
Also, we already have `core.stdc` and that isn't going to change.

**Comment by ibuclaw** [comment]
> has no affect on ABI, which is all we are worried about.

I see this flag as a way to control which templates to instantiate in core.stdcpp.

Mangling abi is mandated by another mechanism, though you are free to pretend otherwise in dmd. :-)

**Comment by thewilsonator** [comment]
That was in the context of `c++NN` vs. `gnu++NN` which has no effect in us. `-fabi-version`, I'm not sure what to do about, but going with whatever the default or newest for the corresponding standard version is will probably end up happening.

**Comment by TurkeyMan** [comment]
> Mangling abi is mandated by another mechanism

These flags affect some ABI matters. For instance, GCC started mangling `std::__cxx11::whatever` when C++11 is flagged. I need to detect those 2 cases. Clang has `std::__1`, etc.

**Comment by ibuclaw** [comment]
> These flags affect some ABI matters. For instance, GCC started mangling std::__cxx11::whatever when C++11 is flagged. I need to detect those 2 cases. Clang has std::__1, etc.

Yep, that is all done in libstdc++ though. The original special std substitutions are unchanged in the compiler.

**Pull request merged**